modules/graceful/manager.go

```diff
@@ -158,37 +171,54 @@ func (g *gracefulManager) doShutdown() {
 		g.doHammerTime(0)
 		<-time.After(1 * time.Second)
 		g.doTerminate()
+		g.WaitForTerminate()
+		g.lock.Lock()
+		if g.done == nil {
+			g.done = make(chan struct{})
+		}
+		close(g.done)
+		g.lock.Unlock()
 	}()
 }
 
-func (g *gracefulManager) doHammerTime(d time.Duration) {
+func (g *Manager) doHammerTime(d time.Duration) {
 	time.Sleep(d)
+	g.lock.Lock()
+	if g.hammer == nil {
+		g.hammer = make(chan struct{})
+	}
+	g.lock.Unlock()
 	select {
 	case <-g.hammer:
 	default:
 		log.Warn("Setting Hammer condition")
+		g.lock.Lock()
```

Comment:
For what I understand about channels (little) it seems to me that this `Lock()` should be placed before the select. Neither way should create a deadlock, but only one goroutine will print "Setting Hammer condition".

Reply:
hmm... the syncing here is wrong. It should be:

```go
func (g *Manager) doHammerTime(d time.Duration) {
	time.Sleep(d)
	g.lock.Lock()
	if g.hammer == nil {
		g.hammer = make(chan struct{})
	}
	select {
	case <-g.hammer:
	default:
		log.Warn("Setting Hammer condition")
		close(g.hammer)
	}
	g.lock.Unlock()
}
```